In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
#  YOUR CODE GOES HERE
columns_to_drop = ['EIN','NAME']
application_df = application_df.drop(columns=columns_to_drop)
# Print the resulting DataFrame
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [4]:
# Determine the number of unique values in each column.
#  YOUR CODE GOES HERE
unique_value_counts = application_df.nunique()
unique_value_counts

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [5]:
# Look at APPLICATION_TYPE value counts for binning
#  YOUR CODE GOES HERE
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
application_type_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#  YOUR CODE GOES HERE
cutoff_value = 10
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
application_types_to_replace = list(application_type_counts[application_type_counts < cutoff_value].index)


# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
T13         66
T12         27
T2          16
Other       11
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Look at CLASSIFICATION value counts for binning
#  YOUR CODE GOES HERE
classification_type_counts = application_df['CLASSIFICATION'].value_counts()
classification_type_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [8]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#  YOUR CODE GOES HERE
classification_type_counts = classification_type_counts[classification_type_counts > 1]
classification_type_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [9]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
#  YOUR CODE GOES HERE
cutoff_value = 100
classification_type_counts = application_df['CLASSIFICATION'].value_counts()

classifications_to_replace = list(classification_type_counts[classification_type_counts < cutoff_value].index)
# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64

In [10]:
# Convert categorical data to numeric with `pd.get_dummies`
#  YOUR CODE GOES HERE
application_df_encoded = pd.get_dummies(application_df)
application_df_encoded

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0           1      5000              1                       0   
1           1    108590              1                       0   
2           1      5000              0                       0   
3           1      6692              1                       0   
4           1    142590              1                       0   
...       ...       ...            ...                     ...   
34294       1      5000              0                       0   
34295       1      5000              0                       0   
34296       1      5000              0                       0   
34297       1      5000              1                       0   
34298       1  36500179              0                       0   

       APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                         1                     0                     0   
1                         0                     0                     0   
2                         0                     0                     0   
3                         0                     0                     0   
4                         0                     0                     0   
...                     ...                   ...                   ...   
34294                     0                     0                     0   
34295                     0                     0                     0   
34296                     0                     0                     0   
34297                     0                     0                     0   
34298                     0                     0                     0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T3  ...  \
0                         0                    0                    0  ...   
1                         0                    0                    1  ...   
2                         0                    0                    0  ...   
3                         0                    0                    1  ...   
4                         0                    0                    1  ...   
...                     ...                  ...                  ...  ...   
34294                     0                    0                    0  ...   
34295                     0                    0                    0  ...   
34296                     0                    0                    1  ...   
34297                     0                    0                    0  ...   
34298                     0                    0                    1  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                      0                       0                         0   
1                      1                       0                         0   
2                      0                       0                         0   
3                      0                       1                         0   
4                      0                       0                         1   
...                  ...                     ...                       ...   
34294                  0                       0                         0   
34295                  0                       0                         0   
34296                  0                       0                         0   
34297                  0                       0                         0   
34298                  0                       0                         0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                       0                 0                       0   
1                       0                 0                       0   
2                       0                 0                       0   
3                       0                 0                       0   
4                       0                 0                       0   
...          

In [11]:
# Split our preprocessed data into our features and target arrays
#  YOUR CODE GOES HERE
y = application_df_encoded['IS_SUCCESSFUL'].values
X = application_df_encoded.drop(columns='IS_SUCCESSFUL').values
# Split the preprocessed data into a training and testing dataset
#  YOUR CODE GOES HERE
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [12]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [13]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE

number_input_features = len(X_train[0])
hidden_nodes_layer1 = 80
hidden_nodes_layer2 = 30

nn = tf.keras.models.Sequential()

# First hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, activation="relu", input_dim=number_input_features))
# Second hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))
# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                4320      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 1)                 31        
                                                                 
Total params: 6,781
Trainable params: 6,781
Non-trainable params: 0
_________________________________________________________________


In [14]:
# Compile the model
#  YOUR CODE GOES HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [16]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn.fit(X_train_scaled, y_train, epochs=500)

Epoch 1/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5299 - accuracy: 0.7421
Epoch 2/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5300 - accuracy: 0.7422
Epoch 3/500
804/804 [==============================] - 2s 2ms/step - loss: 0.5295 - accuracy: 0.7430
Epoch 4/500
804/804 [==============================] - 2s 3ms/step - loss: 0.5300 - accuracy: 0.7421
Epoch 5/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5303 - accuracy: 0.7420
Epoch 6/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5299 - accuracy: 0.7425
Epoch 7/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5292 - accuracy: 0.7423
Epoch 8/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5293 - accuracy: 0.7428
Epoch 9/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5297 - accuracy: 0.7429
Epoch 10/500
804/804 [==============================] - 1s 2ms/step - loss: 0.5299 - accura

In [17]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.6105 - accuracy: 0.7256 - 643ms/epoch - 2ms/step
Loss: 0.610543429851532, Accuracy: 0.7255976796150208


In [18]:
from tensorflow.keras.models import load_model
nn.save('nnmodel.h5')

**Overview of the analysis:**

The goal of the analysis is to help the nonprofit foundation Alphabet Soup select the applicants for funding with the best chance of success in their ventures. A binary classifier was created in order to predict whether applicants would be successful if funded by Alphabet Soup.

There are more than 34,000 organizations that have received funding from Alphabet Soup over the years, and we are looking into that dataset.  The columns that capture metadata about each organization are as follows:

* **EIN** and **NAME**&mdash;Identification columns
* **APPLICATION_TYPE**&mdash;Alphabet Soup application type
* **AFFILIATION**&mdash;Affiliated sector of industry
* **CLASSIFICATION**&mdash;Government organization classification
* **USE_CASE**&mdash;Use case for funding
* **ORGANIZATION**&mdash;Organization type
* **STATUS**&mdash;Active status
* **INCOME_AMT**&mdash;Income classification
* **SPECIAL_CONSIDERATIONS**&mdash;Special considerations for application
* **ASK_AMT**&mdash;Funding amount requested
* **IS_SUCCESSFUL**&mdash;Was the money used effectively

The steps in this analysis were data preprocessing and compiling,training and evaluating the model.

* Data preprocessing
  * we took the column "IS_SUCCESSFUL" as the target - it is key data that tells us whether the money was used effectively
  * the features of the model are: application type, affiliation, classification, use_case, organization, status, income_amt, special_considerations and ask_amt-funding amount requested
  * EIN and NAME were removed as columns-variables since as identification columns they do not contribute to the quality of the analysis

* Compiling, Training, and Evaluating the Model
  * The initial neural network model had two layers with 80 neurons in the first layer and 30 neurons in the second layer, with 100 epochs. The accuracy achieved was 0.0.7258309125900269. The goal was to optimize the model to achieve accuracy of at least 75%
  * Steps taken were:
  1) reduction of the no of bins in preprocessing stage for the application type - result was acc of 72.4%.
  2) increase of the no of bins in preprocessing stage for the application type AND for the classification type by changing cutoff value - result was 72.59% accuracy
  3) increasing number of neurons in layer 1 from 80 to 90 - result was 72.54% accuracy
  4) increasing no of epochs from 100 to 500 - result was 72.56% accuracy

Summary - Overall, the steps taken to optimize the model did not achieve significantly higher accuracy. There was also an option to reduce the number of columns - variables to streamline the model, however, I decided not to take the steps since I am not entirely familiar with the meaning of codes and types in the dataset and I did not want any important information to be lost. In further optimization steps the model can be changed from Relu to another model.
